Updates "SameSite cookies explainer" and adds "SameSite cookies recipes" (#1367)
GoogleChrome/web.dev: src/site/content/en/blog/samesite-cookie-recipes/index.md

---
title: SameSite cookie recipes
subhead:
  Update your site's cookies to prepare for the upcoming changes to the SameSite
  attribute's behavior.
authors:
  - rowan_m
date: 2019-10-15
hero: cookie-hero.jpg
description: |
  With the introduction of the new SameSite=None attribute value, sites can now
  explicitly mark their cookies for cross-site usage. Browsers are moving to
  make cookies without a SameSite attribute act as first-party by default, a
  safer and more privacy preserving option than the current open behavior.
  Learn how to mark up your cookies to ensure your first-party and third-party
  cookies continue to work once this change comes into effect.
tags:
  - post
  - security
  - cookies
---

{% Aside %}

If you need a refresher on cookies and the `SameSite` attribute, head over to

> the earlier companion article,

Reviewer: You can delete this whole line without losing meaning.

Author: I think what I'm looking for here is a standard widget for a series of posts. My concern is that just a link to the article doesn't explain why you would follow it. I feel like I need to provide some kind of context.

Do you think the following would work in the respective articles?

- *(In part 1)* For implementation advice on `SameSite=None`, see part 2: "`SameSite` cookie recipes"
- *(In part 2)* For how cookies and `SameSite` work, see part 1: "`SameSite` cookies explained"

Reviewer: I was only suggesting deleting 'the earlier companion article'. The stuff before it is good.

Author: I've gone with some compact part 1  / part 2 introductions,

["SameSite cookies explained"](/samesite-cookies-explained).

{% endAside %}

> To recap from before: [Chrome](https://www.chromium.org/updates/same-site),

Reviewer: Delete 'from before'. 

The reason for this is the same as the previous suggestion, [Strunk and White](https://www.amazon.com/Elements-Style-Fourth-William-Strunk/dp/020530902X/ref=pd_sbs_14_t_0/140-9010431-1453962?_encoding=UTF8&pd_rd_i=020530902X&pd_rd_r=f53e256d-a283-4b76-997c-5c58c53af430&pd_rd_w=1qObN&pd_rd_wg=sszWF&pf_rd_p=5cfcfe89-300f-47d2-b1ad-a4e27203a02a&pf_rd_r=JWVHE4V0RS2BJEFACQVD&psc=1&refRID=JWVHE4V0RS2BJEFACQVD) rule II 17, "Omit needless words."

Reviewer: By the way, there's another rule, V 19, "Do not take shortcuts at the cost of clarity." It doesn't seem to be a problem here. I mention it because much of technical writing is about finding the happy path between the two rules.

Author: Recap removed.

[Firefox](https://groups.google.com/d/msg/mozilla.dev.platform/nx2uP0CzA9k/BNVPWDHsAQAJ),
[Edge](https://textslashplain.com/2019/09/30/same-site-cookies-by-default/), and
others will be changing their default behavior so that:

1. cookies without a `SameSite` attribute will be treated as `SameSite=Lax`, or

>    restricted to first-party contexts

Reviewer: @robdodson @kaycebasques This is a case where our practice of presenting everything as shiny may not be serving the internet well. There needs to be some attention to this at least in the description, because this is going to break the web.

Reviewer: I didn't at first realize there were two articles here. This comment applies to the announcement article but not this one.

Reviewer 2: @jpmedley I updated the title of this PR to clarify that there are 2 posts. Regarding your first comment, can you provide suggestions on what needs to change and what we should change it to?

Reviewer: Maybe "Same site cookies are coming. Make sure you're ready?"

I don't know how to be strong enough without being negative. We have to say, "We're about to break your site", without actually saying it.

Author: I've tried to make this more explicit in the latest edits.

2. cookies for cross-site or third-party usage must specify
   `SameSite=None; Secure`

## Use cases for cross-site or third-party cookies

There are a number of common use cases and patterns where cookies need to be
sent in a third-party context. If you provide or depend on one of these use
cases then you will need to ensure that either you or the provider are updating
their cookies to ensure the service continues to function correctly.

### Content within an `<iframe>`

When content is displayed within an `<iframe>` this qualifies as a third-party

> context. Standard use cases here are:

Reviewer: Easily shortened to, "Content in an `<iframe>` is in a third-party context.

Author: Shortened!

- embedded content shared from other sites, such as videos, maps, code samples,
  social posts
- widgets from external services such as payments, calendars, booking /
  reservation services

> - widgets such as social buttons or anti-fraud services that create less obvious

Reviewer: The last few items in this list are missing their commas.

Author: Added.

  `<iframes>`

Cookies may be used here to personalize content if the current user has an
existing account on the remote site, maintain a session state, store general

> preferences, enable usage statistics, and more.

Reviewer: Suggested rewrite: "Cookies may be used here to, among other things, maintain session state, store general preferences, enable statistics, or personalize content for users with existing accounts."

I move personalized content to the end because it is so long that I didn't initially read 'maintain session state' as the second item in a list. I was scratching my head for a moment. Replacing 'and more' with 'among other things' is simply personal preference.

Author: Updated.

<figure class="w-figure  w-figure--center">
  <img src="iframe.png"
      alt="Diagram of a browser window where the URL of embedded content does
        not match the URL of the page."
      style="max-width: 35vw;">
  <figcaption class="w-figcaption">
    If the embedded content doesn't come from the same site as the top-level
    browsing context, it's third-party content.
  </figcaption>
</figure>

Additionally, as the web is inherently composable, `<iframes>` are used to embed
content that is also viewed in a top-level or first-party context. Any cookies
used by that site will be considered as third-party cookies when the site is
displayed within the frame. If you're creating sites that you intend to be

> easily composed that also rely on cookies to function, you will also need to

Reviewer: What is an easily composed site?

Author: Switched to "embedded" to hopefully make this more obvious.

ensure those are marked for cross-site usage or that you can gracefully fallback
without them.

### "Unsafe" requests across sites

While "unsafe" may sound slighly concerning here, this refers to any request

> that may be intended to change state and on the web that's primarily POST

Reviewer: This sentence is already a mouthful. I'd start a new sentence at 'On the web'.

Author: Split sentence.

requests. Cookies marked as `SameSite=Lax` will be sent on safe top-level
navigations, e.g. clicking a link to go to a different site. However something
like a `<form>` submission via POST to a different site would not include
cookies.

<figure class="w-figure  w-figure--center">
  <img src="safe-navigation.png"
      alt="Diagram of a request moving from one page to another."
      style="max-width: 35vw;">
  <figcaption class="w-figcaption">
    If the incoming request uses a "safe" method then the cookies will be sent.
  </figcaption>
</figure>

This pattern is used for sites that may redirect the user out to a remote
service to perform some operation before returning, for example redirecting to a
third-party identity provider. Before the user leaves the site, a cookie is set
containing a single use token with the expectation that this token can be

> checked on the returning request to mitigate CSRF attacks. If that returning

Reviewer: CSRF: spell out cross-site request forgery attacks and link to an article or description. Here's the article they're linking to on the Chromium blog. https://cheatsheetseries.owasp.org/cheatsheets/Cross-Site_Request_Forgery_Prevention_Cheat_Sheet.html

Author: Linked.

request comes via POST then it will be necessary to mark the cookies as
`SameSite=None; Secure`.

### Remote resources

> Any remote resource on a page may be relying on cookies to be sent with that

Reviewer: Change 'that' to 'a'. 'That' is a pronoun which requires an antecedent.

Author: Switched.

> request, from `<img>` tags, `<script>` tags, and so on. Common use cases here

Reviewer: Delete 'here'.

Author: Removed.

include tracking pixels and personalizing content.

This also applies to
[`fetch` requests](https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API/Using_Fetch#Sending_a_request_with_credentials_included)
from the page. If you see the `credentials: 'include'` option in the

> configuration this is a good indication that cookies may well be expected on

Reviewer: "If you see the `credentials: 'include'` option in the configuration..."

Change to:

"If `fetch()` is called with the `credentials: 'include'` option..."

It states what you're referring to more clearly.

Author: Replaced.

> those requests. Those cookies will need to be appropriate marked to be included.

Reviewer: appropriate --> appropriately

Author: Fixed.

## How to implement `SameSite` today

> For cookies where you know they are only needed in a first-party context you

Reviewer: Delete 'where you know they are'. It doesn't add meaning.

Author: Deleted.

should ideally mark them as `SameSite=Lax` or `SameSite=Strict` depending on
your needs. You can also choose to do nothing and just allow the browser to
enforce its default, but this comes with the risk of inconsistent behavior
across browsers and potential console warnings for each cookie.

```text
Set-Cookie: first_party_var=value; SameSite=Lax
```

> For cookies needed in a third-party context, you will need to ensure thet are

Reviewer: Delete 'you will need to'.

Author: Removed.

marked as `SameSite=None; Secure`. Note that you need both attributes together.
If you just specify `None` without `Secure` the cookie will be rejected. There

> are some mutually incompatible differences in browser implementations here

Reviewer: Delete 'here'.

Author: Deleted.

though, so you may need to use some of the mitigating strategies described in
["Handling incompatible clients"](#handling-incompatible-clients) below.

```text
Set-Cookie: third_party_var=value; SameSite=None; Secure
```

### Identifying cookie usage

> As of Chrome 77, you will be able to see warnings in the console cross-site

Reviewer: Delete 'will be able to'.

Missing a word at the end: 'in the console **for** cross-site'.

Author: Fixed.

> can review cookies in developer tools under Application>Storage>Cookies and
> see more details at
> [https://www.chromestatus.com/feature/5633521622188032](https://www.chromestatus.com/feature/5633521622188032)."

Each of these warnings will contain the cookie domain. If you're responsible for
that domain, then you will need to update the cookies. Otherwise, you may need
to contact the owner of the site or service responsible for that cookie to
ensure they making the necessary changes. The warnings themselves do not affect
the functionality of the site, this is purely to inform developers of the
upcoming changes.

### Handling incompatible clients

As these changes to include `None` and update default behavior are still
relatively new, there are inconsistencies amongst browsers as to how these
changes are handled. You can refer to the
[updates page on chromium.org](https://www.chromium.org/updates/same-site/incompatible-clients)
for the issues currently known, however it's not possible to say if this is
exhaustive. While this is not ideal, there are workarounds you can employ during
this transitionary phase. The general rule though is to treat incompatible

> clients as the special case - do not create an exception for browsers

Reviewer: 'Do not create...' is the start of a new sentence.

Author: Split.

---
title: SameSite cookie recipes
subhead:
  Update your site's cookies to prepare for the upcoming changes to the SameSite
  attribute's behavior.
authors:
  - rowan_m
date: 2019-10-24
hero: cookie-hero.jpg
description: |
  With the introduction of the new SameSite=None attribute value, sites can now
  explicitly mark their cookies for cross-site usage. Browsers are moving to
  make cookies without a SameSite attribute act as first-party by default, a
  safer and more privacy preserving option than the current open behavior.
  Learn how to mark up your cookies to ensure your first-party and third-party
  cookies continue to work once this change comes into effect.
tags:
  - post
  - security
  - cookies
---

{% Aside %}

For how cookies and `SameSite` work, see part 1:

> ["SameSite cookies explained"](/samesite-cookies-explained).

Reviewer 2: ```suggestion
[SameSite cookies explained](/samesite-cookies-explained).
```

{% endAside %}

[Chrome](https://www.chromium.org/updates/same-site),
[Firefox](https://groups.google.com/d/msg/mozilla.dev.platform/nx2uP0CzA9k/BNVPWDHsAQAJ),
[Edge](https://textslashplain.com/2019/09/30/same-site-cookies-by-default/), and
others will be changing their default behavior in line with the IETF proposal,

> ["Incrementally Better Cookies"](https://tools.ietf.org/html/draft-west-cookie-incrementalism-00)

Reviewer 2: ```suggestion
[Incrementally Better Cookies](https://tools.ietf.org/html/draft-west-cookie-incrementalism-00)
```

so that:

1. cookies without a `SameSite` attribute will be treated as `SameSite=Lax`,
   meaning the default behavior will be to restrict cookies to first party
   contexts **only**.
2. cookies for cross-site usage **must** specify `SameSite=None; Secure` to
   enable inclusion in third party context.

This will become the default behavior in Chrome 80, planned for a stable release

> in February, 2020. If you currently provide cookies that are intended for

Reviewer 2: ```suggestion
in February 2020. If you currently provide cookies that are intended for
```

this is a good indication that cookies may well be expected on those requests.
Those cookies will need to be appropriately marked to be included.

### Content within a WebView

A WebView in a native app is powered by a browser and the same restrictions and
issues will apply. In Android, if the WebView is powered by Chrome, the default
behavior will change with the release of Chrome 80. Additionally, Android allows
native apps to set cookies directly via the
[CookieManager API](https://developer.android.com/reference/android/webkit/CookieManager).
As with cookies set via headers or JavaScript, these must include
`SameSite=None; Secure` if they are intended for cross-site use.

## How to implement `SameSite` today

For cookies where they are only needed in a first-party context you
should ideally mark them as `SameSite=Lax` or `SameSite=Strict` depending on
your needs. You can also choose to do nothing and just allow the browser to
enforce its default, but this comes with the risk of inconsistent behavior
across browsers and potential console warnings for each cookie.

```text
Set-Cookie: first_party_var=value; SameSite=Lax
```

> For cookies needed in a third-party context, you will need to ensure thet are

Reviewer 2: ```suggestion
For cookies needed in a third-party context, you will need to ensure they are
```

marked as `SameSite=None; Secure`. Note that you need both attributes together.
If you just specify `None` without `Secure` the cookie will be rejected. There
are some mutually incompatible differences in browser implementations though, so
you may need to use some of the mitigating strategies described in

> ["Handling incompatible clients"](#handling-incompatible-clients) below.

Reviewer 2: ```suggestion
[Handling incompatible clients](#handling-incompatible-clients) below.
```

And for `None` without `Secure`, you will see:

> "A cookie associated with a resource at (cookie domain) was set with
> `SameSite=None` but without `Secure`. A future release of Chrome will only
> deliver cookies marked `SameSite=None` if they are also marked `Secure`. You
> can review cookies in developer tools under Application>Storage>Cookies and
> see more details at
> [https://www.chromestatus.com/feature/5633521622188032](https://www.chromestatus.com/feature/5633521622188032)."

Each of these warnings will contain the cookie domain. If you're responsible for
that domain, then you will need to update the cookies. Otherwise, you may need
to contact the owner of the site or service responsible for that cookie to
ensure they making the necessary changes. The warnings themselves do not affect
the functionality of the site, this is purely to inform developers of the
upcoming changes.

<figure class="w-figure  w-figure--center">
  <img src="samesite-devtools.png"
      alt="Chrome Developer Tools showing the overview of cookies for a site"
      style="max-width: 40vw;">
  <figcaption class="w-figcaption">

>     Chrome Developer Tools shows the values for the <tt>SameSite</tt> and

Reviewer 2: ```suggestion
    Chrome DevTools shows the values for the <tt>SameSite</tt> and
```
